hw/net/rocker/rocker_p4_rmt.c

```diff
@@ -0,0 +1,216 @@
+                                    char *buf, unsigned short cmd,
+                                    RockerTlv **tlvs);
+
+unsigned char *
+rocker_p4_iov_to_buffer(const struct iovec *iov, int iovcnt)
+{
+    // convert iov to a single contiguous buffer for p4 model
+    int len = 0;
+    unsigned char *buf = NULL;
+
+    len = iov_size(iov, iovcnt);
+    DPRINTF("convert %d iovs to a buffer of len %d\n", iovcnt, len);
+    buf = malloc(len);
+    if (buf == NULL) {
+        DPRINTF("Cannot allocate buffer\n");
+        return buf;
+    }
+    iov_to_buf(iov, iovcnt, 0, buf, len);
+    return buf;
+}
+
+
+void rocker_p4_rmt_tx (int eg_port1, void *pkt, int len, int ig_port1, 
+                            void *cookie)
+{
+    struct iovec iov;
+    int    iovcnt = 1;
+    struct rocker *rocker;
+    struct world *w;
+    unsigned int fp_ig_port;
+
+    struct p4_rmt_world *p4_rmt = (struct p4_rmt_world *)cookie; // world_private(w);
+    rocker = p4_rmt->rocker;
+    w = p4_rmt->world;
+
+    fp_port_from_pport((unsigned int)ig_port1, &fp_ig_port);
+
+    DPRINTF("rocker_p4_rmt_tx to port %d, len %d.... \n", eg_port1, len);
+    if (eg_port1 == 0) {
+        DPRINTF("rocker_p4_rmt_tx DROP .... \n");
+        free (pkt);
+        return;
+    }
+    // convert the packet to iov
+    iov.iov_len = len;
+    iov.iov_base = pkt;
+    if (rocker == NULL) {
+        DPRINTF("\"rocker\" not found\n");
+        return;
+    }
+    if (w == NULL) {
+        DPRINTF("P4-RMT world not found\n");
+        return;
+    }
+    // check for sup port
+    if (p4_rmt->is_cpu_port(eg_port1)) {
+        DPRINTF("rocker_p4_rmt_tx to CPU from ingress port %d, len %d.... \n", 
+                    ig_port1, len);
+        rx_produce(w, ig_port1, &iov, iovcnt);
+        return;
+    }
+    // call qemu_send to send out on switch port
+    rocker_port_eg(rocker, eg_port1, &iov, iovcnt);
+    return;
+}
+
+static int rocker_p4_rmt_table_cmd (struct p4_rmt_world *p4_rmt, 
+                            struct desc_info *info,
+                            char *buf, unsigned short cmd,
+                            RockerTlv **tlvs)
+{
+    unsigned int    table_id;
+    unsigned char   *entry;
+    if (tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID] == NULL ||
+        tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ENTRY] == NULL) {
+        DPRINTF("Missing TLVs\n");
+        return 0;
+    }
+        
+    table_id = rocker_tlv_get_u32(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ID]);
+    entry = rocker_tlv_data(tlvs[ROCKER_TLV_P4_RMT_INFO_TABLE_ENTRY]);
+
+    DPRINTF("p4_rmt_table_cmd for table %d\n", table_id);
+
+    if (!p4_rmt->is_table_valid(table_id)) {
+        DPRINTF("p4_rmt_table_cmd for invalide table %d\n", table_id);
+        return 0;
+    }
+
+    switch (cmd) {
+        case ROCKER_TLV_CMD_TYPE_P4_RMT_TABLE_ENTRY_ADD:
+            // check if fn is provided
+            if (p4_rmt->table_ops[table_id].add) {
+                p4_rmt->table_ops[table_id].add(entry);
```

Comment:
If it's generated code, then can a TLV format be generated?  I know that brings rocker's TLV packing into P4 tools, which is probably not what you want.  I think a blob is not going to be the correct answer, but I'm not sure where to slice this.  This is something we should revisit after all the other review work is done.